### PR TITLE
#1906 Implemented issue requirements.

### LINKED ISSUE #1906: Optimistic locking: automatic conflict resolution should be optional
### Description
By default, optimistic locking is applied to all persistent entities. Conflicts are recognised based on the versioning mechanism. And by default, automatic conflict resolution takes place, which may or may not succeed.

In practice, there are situations where automatic conflict resolution needs to be avoided, and instead a conflict error should be thrown, if the version-based mechanism identified a conflict.

All persistent entities have annotation `MapEntityTo`, which is a suitable place to control this behaviour. It is proposed to introduce attribute `autoConflictResolution` set to `true` by default, and inspect the value of this attribute in `PersistentEntitySaver.canResolveConflict` to return `false`, if `autoConflictResolution == false`, without attempting to resolve a source of conflict.

Entities, for which conflict recognition should strictly be based on `version`, should be annotated `MapEntityTo(autoConflictResolution = false)`.

### Expected outcome
Ability to control applicability of the automatic conflict resolution at the level of domain entities.

### PR DESCRIPTION
**Pull request by 01es**
Resolve #1906

### CODE REVIEW

File: platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java [hunk 1]
```diff
@@ -51,11 +53,10 @@ public void conflicting_changes_should_have_prevent_entity_saving() {
         } catch (final EntityCompanionException ex) {
             assertEquals(format("%s Tg Category [Cat1] could not be saved.", ERR_COULD_NOT_RESOLVE_CONFLICTING_CHANGES), ex.getMessage());
         }
-
     }
 
     @Test
-    public void concurrent_setting_of_activatable_property_should_not_lead_to_conflict_resolution_errors() {
+    public void concurrent_modification_of_refCount_due_to_new_activatable_dependencies_does_not_lead_to_conflict_resolution_errors() {
```
Comment (elvin-fms) [inline platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java:59]: Could this have an undesired side-effect on activatable entities with activatable dependencies if `autoConflictResolution` is turned off for them?
No PO lines in TG Air currently have activatable dependencies, but what if `autoConflictResolution` is turned off for some other entities, which do?

File: platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java [hunk 2]
```diff
@@ -28,7 +30,7 @@ public void non_conflicting_changes_should_have_been_resolved() {
     }
 
     @Test
-    public void identical_conflicting_changes_should_have_been_resolved() {
+    public void identical_concurrent_changes_should_have_been_resolved() {
```
[the remark this answers is not in the document]
Comment (01es) [reply platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java:33]: So should we all!

[the remark this answers is not in the document]
Comment (01es) [reply platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java:33]: There was an attempt to update old test names, but this one must have slipped.

File: platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java (hunk 1 above)
Comment (01es) [reply platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java:59]: It should not and that was verified. A test could be provided.

Comment (elvin-fms) [reply platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java:59]: Could be a good idea for complete certainty, unless #1745 is going to be implemented really soon.

File: platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java [hunk 3]
```diff
@@ -64,15 +66,62 @@ public void concurrent_setting_of_activatable_property_should_not_lead_to_confli
         assertEquals(Integer.valueOf(3), co(TgCategory.class).findByKey("Cat1").getRefCount());
     }
 
+    @Test
+    public void concurrent_modification_of_refCount_due_to_new_activatable_dependencies_does_not_lead_to_conflict_resolution_errors_for_entities_with_not_auto_conflict_resolution() {
+        final TgBogieClass bc = co(TgBogieClass.class).findByKey("BC1");
+        assertEquals(Integer.valueOf(0), bc.getRefCount());
+
+        save(new_(TgBogie.class, "Bogie2").setBogieClass(bc));
+        save(new_(TgBogie.class, "Bogie3").setBogieClass(bc));
```
Comment (elvin-fms) [inline platform-dao/src/test/java/ua/com/fielden/platform/entity/AutomaticConflictResolutionTest.java:75]: Pardon me, but I don't really see concurrency here, it all looks very sequential.

I appreciate that all other tests here have two sequential invocations of `save()`, but this test is different. Other tests retrieve two instrumented instances of the same entity and then operate with them. In this test, I suppose `refCount` is not incremented on the non-instrumented instance of `bc` assigned to property `bogieClass`. Instead I suspect a fresh instrumented instance is retrieved during each sequential `save()`.

Actually, the same probably applies to `concurrent_modification_of_refCount_due_to_new_activatable_dependencies_does_not_lead_to_conflict_resolution_errors()`

**Review by elvin-fms — CHANGES_REQUESTED** [review]

**Review by elvin-fms — APPROVED** [review]

**Pull request merged**